### PR TITLE
feat: add TextTokenCount for type Message

### PR DESCRIPTION
**Pull request by Sherlocksuper**

### CODE REVIEW

File: object/message.go [hunk 1]
```diff
@@ -156,6 +157,12 @@ func UpdateMessage(id string, message *Message) (bool, error) {
 		return false, nil
 	}
 
+	size, err := GetMessageTextTokenCount(message.ModelProvider, message.Text)
```
Comment (hsluoyz) [inline object/message.go]: if TextTokenCount is not 0 and text is not changed, don't run

File: object/message.go [hunk 2]
```diff
@@ -215,6 +222,11 @@ func RefineMessageFiles(message *Message, origin string) error {
 }
 
 func AddMessage(message *Message) (bool, error) {
+	size, err := GetMessageTextTokenCount(message.ModelProvider, message.Text)
```
Comment (hsluoyz) [inline object/message.go]: Will message.ModelProvider be non-empty for all cases? Not really I think

File: object/message.go [hunk 3]
```diff
@@ -334,3 +346,14 @@ func GetPaginationMessage(owner string, offset, limit int, field, value, sortFie
 
 	return messages, nil
 }
+
+func GetMessageTextTokenCount(modelName string, text string) (int, error) {
+	tokenCount, err := model.GetTokenSize(modelName, text)
```
Comment (hsluoyz) [inline object/message.go:353]: Handle modelName is empty

File: object/message.go [hunk 4]
```diff
@@ -334,3 +346,14 @@ func GetPaginationMessage(owner string, offset, limit int, field, value, sortFie
...
+	if err != nil {
+		tokenCount, err = model.GetTokenSize("gpt-3.5-turbo", text)
```
Comment (hsluoyz) [inline object/message.go:355]: When enter this if?

File: web/src/MessageListPage.js [hunk 5]
```diff
@@ -200,6 +200,14 @@ class MessageListPage extends BaseListPage {
         sorter: (a, b) => a.tokenCount - b.tokenCount,
         // ...this.getColumnSearchProps("tokenCount"),
       },
+      {
+        title: i18next.t("chat:Text token count"),
+        dataIndex: "textTokenCount",
+        key: "textTokenCount",
+        width: "90px",
+        sorter: (a, b) => a.tokenCount - b.tokenCount,
```
Comment (hsluoyz) [inline web/src/MessageListPage.js]: textTokenCount

File: object/message.go [hunk 6]
```diff
@@ -334,3 +346,14 @@ func GetPaginationMessage(owner string, offset, limit int, field, value, sortFie
 
 	return messages, nil
 }
+
+func GetMessageTextTokenCount(modelName string, text string) (int, error) {
```
Comment (hsluoyz) [inline object/message.go]: getMessageTextTokenCount

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: object/message.go (hunk 3 above)
Comment (Sherlocksuper) [reply object/message.go:353]: err!＝nil handle this

File: object/message.go (hunk 4 above)
Comment (Sherlocksuper) [reply object/message.go:355]: modelname is empty or not supported by tiktoken

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**